File: php/sweethearts/HighSchoolSweetheart.php
<?php declare(strict_types=1);

class HighSchoolSweetheart
{
    public function firstLetter(string $name): string
    {
        return trim($name)[0];
    }

    public function initial(string $name): string
    {
        $first = strtoupper($this->firstLetter($name));
        return "$first.";
    }

    public function initials(string $name): string
    {
        return implode(' ', array_map($this->initial(...), explode(' ', $name)));
    }

    public function pair(string $sweetheart_a, string $sweetheart_b): string
    {
        $initials_a = $this->initials($sweetheart_a);
        $initials_b = $this->initials($sweetheart_b);

        return <<<HEART
     ******       ******
   **      **   **      **
 **         ** **         **
**            *            **
**                         **
**     $initials_a  +  $initials_b     **
 **                       **
   **                   **
     **               **
       **           **
         **       **
           **   **
             ***
              *
HEART;
    }
}

File: elixir/high-school-sweetheart/lib/high_school_sweetheart.ex
defmodule HighSchoolSweetheart do
  def first_letter(name) do
    String.trim(name)
    |> String.at(0)
  end

  def initial(name) do
    first_letter(name)
    |> String.capitalize()
    |> Kernel.<>(".")
  end

  def initials(full_name) do
    [first, last] = String.split(full_name, " ")
    "#{initial(first)} #{initial(last)}"
  end

  def pair(full_name1, full_name2) do
  """
       ******       ******
     **      **   **      **
   **         ** **         **
  **            *            **
  **                         **
  **     #{initials(full_name1)}  +  #{initials(full_name2)}     **
   **                       **
     **                   **
       **               **
         **           **
           **       **
             **   **
               ***
                *
  """
  end
end
